### PR TITLE
send back gas to relayer

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/smartContract/process.go [hunk 1]
```diff
@@ -600,37 +603,29 @@ func (sc *scProcessor) processSCRForSender(
 	txHash []byte,
 	vmInput *vmcommon.ContractCallInput,
 	vmOutput *vmcommon.VMOutput,
-	acntSnd state.UserAccountHandler,
 	lockedGas uint64,
-) (*smartContractResult.SmartContractResult, error) {
+) (*smartContractResult.SmartContractResult, *smartContractResult.SmartContractResult, error) {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:607]: Although this works, returning 2 consecutive elements of the same type is always prone to faults (caused by switching variables)

File: process/smartContract/process.go [hunk 2]
```diff
@@ -835,6 +830,9 @@ func (sc *scProcessor) processForRelayerWhenError(
 	if !isRelayed {
 		return nil
 	}
+	if relayedSCR.Value.Cmp(zero) == 0 {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:833]: test relayedSCR.Value for nil?

File: process/smartContract/process.go [hunk 3]
```diff
@@ -1327,12 +1347,12 @@ func (sc *scProcessor) createSmartContractResults(
...
 // createSCRForSender(vmOutput, tx, txHash, acntSnd)
 // give back the user the unused gas money
-func (sc *scProcessor) createSCRForSender(
+func (sc *scProcessor) createSCRForSenderAndRelayer(
 	vmOutput *vmcommon.VMOutput,
 	tx data.TransactionHandler,
 	txHash []byte,
 	callType vmcommon.CallType,
-) *smartContractResult.SmartContractResult {
+) (*smartContractResult.SmartContractResult, *smartContractResult.SmartContractResult) {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1355]: same thing about multiple, same-type variables.

File: process/smartContract/process.go [hunk 4]
```diff
@@ -1115,6 +1106,29 @@ func (sc *scProcessor) processVMOutput(
...
+	if check.IfNil(userAcc) {
+		return nil
+	}
+
+	err = userAcc.AddToBalance(value)
+	if err != nil {
+		return err
+	}
+
+	err = sc.accounts.SaveAccount(userAcc)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:1124]: done

File: process/smartContract/process.go [hunk 5]
```diff
@@ -1345,17 +1365,36 @@ func (sc *scProcessor) createSCRForSender(
 		rcvAddress = tx.GetRcvAddr()
 	}
 
+	gasRemaining := vmOutput.GasRemaining
+	var refundGasToRelayer *smartContractResult.SmartContractResult
+	relayedSCR, isRelayed := isRelayedTx(tx)
+	if isRelayed && callType != vmcommon.AsynchronousCall && refundErd.Cmp(zero) > 0 {
+		refundGasToRelayer = &smartContractResult.SmartContractResult{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:1372]: done

File: process/smartContract/process.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/process.go:1355]: any better idea ? :)

File: process/smartContract/process.go (hunk 2 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:833]: ok

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**